google/cloud/bigquery/dbapi/cursor.py

```diff
@@ -209,8 +217,14 @@ def executemany(self, operation, seq_of_parameters):
             seq_of_parameters (Union[Sequence[Mapping[str, Any], Sequence[Any]]]):
                 Sequence of many sets of parameter values.
         """
-        for parameters in seq_of_parameters:
-            self.execute(operation, parameters)
+        if seq_of_parameters:
+            formatted_operation, parameter_types = _format_operation(
+                operation, seq_of_parameters[0]
```

Comment:
Could you describe in a comment why we grab just the first one? I guess either we get the types correct on the first go, or we have a problem anyway?

~Also it's a niche case, but it's possible that someone could pass in a generator or something for the sequence of parameters. In which case `[0]` won't work. I think we need to either always loop like we had (and add some logic to only extract the types once) or do some funny business with the low-level iteration function.~

~I probably should have had a test case with this (generator as parameters) when I documented the allowed type as a Sequence.~

Edit: Nevermind, Sequence supports integer-based item access https://docs.python.org/3/glossary.html#term-sequence It's "Iterable" that only supports iteration. We might actually want to support this use case, but probably best to wait until we actually get a customer request for that feature.

Reply:
Comment added.